plutus-contract-exe/src/Language/Plutus/Contract/State.hs

```diff
@@ -0,0 +1,324 @@
+constructors of 'StatefulContract' and 'Record'. 
+
+I did attempt to reduce this by using 'newtype Rec i a = Rec Either (OpenRecord i) (ClosedRecord i, a)' and variations but always ran into trouble. I would be 
+very grateful for suggestions on how to improve this situation.
+
+-}
+
+data StatefulContract a where
+    CMap :: (a' -> a) -> StatefulContract  a' -> StatefulContract  a
+    CAp :: StatefulContract  (a' -> a) -> StatefulContract  a' -> StatefulContract  a
+    CBind :: StatefulContract  a' -> (a' -> StatefulContract  a) -> StatefulContract  a
+
+    CContract :: C.ContractPrompt Maybe a -> StatefulContract  a
+    CJSONCheckpoint :: (Aeson.FromJSON a, Aeson.ToJSON a) => StatefulContract  a -> StatefulContract  a
+
+initialise
+    :: ( MonadWriter Hooks m )
+    => StatefulContract a
+    -> m (Either (OpenRecord Event) (ClosedRecord Event, a))
+initialise = \case
+    CMap f con -> fmap (fmap f) <$> initialise con
+    CAp conL conR -> do
+        l' <- initialise conL
+        r' <- initialise conR
+        case (l', r') of
+            (Left l, Left r)             -> pure $ Left (OpenBoth l r)
+            (Right (l, _), Left r)       -> pure $ Left (OpenRight l r)
+            (Left l, Right (r, _))       -> pure $ Left (OpenLeft l r)
+            (Right (l, f), Right (r, a)) -> pure $ Right (ClosedBin l r, f a)
+    CBind c f -> do
+        l <- initialise c
+        case l of
+            Left l' -> pure $ Left (OpenBind l')
+            Right (l', a) -> do
+                r <- initialise (f a)
+                case r of
+                    Left r'       -> pure $ Left $ OpenRight l' r'
+                    Right (r', b) -> pure $ Right (ClosedBin l' r', b)
+    CContract con -> do
+        r <- runConM mempty con
+        case r of
+            Nothing -> pure $ Left $ OpenLeaf mempty
+            Just a  -> pure $ Right (ClosedLeaf (FinalEvents mempty), a)
+    CJSONCheckpoint con -> do
+        r <- initialise con
+        case r of
+            Left _       -> pure r
+            Right (_, a) -> pure $ Right (jsonLeaf a, a)
+
+checkpoint :: (Aeson.FromJSON a, Aeson.ToJSON a) => StatefulContract  a -> StatefulContract  a
+checkpoint = CJSONCheckpoint
+
+prtty :: StatefulContract  a -> String
+prtty = \case
+    CMap _ c -> "cmap (" ++ prtty c ++ ")"
+    CAp l r -> "cap (" ++ prtty l ++ ") (" ++ prtty r ++ ")"
+    CBind l _ -> "cbind (" ++ prtty l ++  ") f"
+    CContract _ -> "ccontract"
+    CJSONCheckpoint j -> "json(" ++ prtty j ++ ")"
+
+instance Functor StatefulContract where
+    fmap f = \case
+        CMap f' c -> CMap (f . f') c
+        CAp l r     -> CAp (fmap (fmap f) l) r
+        CBind m f'  -> CBind m (fmap f . f')
+
+        CContract con -> CContract (fmap f con)
+        CJSONCheckpoint c -> CMap f (CJSONCheckpoint c)
+
+instance Applicative StatefulContract where
+    pure = CContract . pure
+    (<*>) = CAp
+
+-- TODO: Should we add an `Alt` constructor to `StatefulContract`?
```

Comment:
Don't we need one? I'd have thought that this would impede us from efficiently saving contracts defined via `<|>`, because they'll be lowered to `CContract`s so we'll only be able to serialize them when *all* the branches have reached a checkpoint. Whereas if there was a constructor I think we could do that.

Reply:
Yeah good point.